### ISSUE TITLE
Rule proposal: safe arguments

### ISSUE THREAD

**Issue by zloirock** [eslint/eslint#3504]
Usage `arguments` cause some deoptimizations, but usage ES6 rest arguments everywhere is not possible. I propose add new *disabled by default* rule for notify about [possible problems](https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#3-managing-arguments), as minimum:
* Calling function with `arguments` (exclude `.apply`).
* Reassign defined parameter if function also uses `arguments`.

Possible merge with it [no-caller](http://eslint.org/docs/rules/no-caller) rule.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by ilyavolodin** [comment]
Can you provide some code examples that should be marked as invalid, and how you expect them to look in a valid case please?

**Comment by zloirock** [comment]
```js
// invalid:
foo(arguments);
var args = Array.prototype.slice.call(arguments);

// valid:
baz.apply(null, arguments);

var args = new Array(arguments.length);
for(var i = 0; i < args.length; ++i) {
  args[i] = arguments[i];
}

// invalid:
function f(a){
  if(arguments.length < 1)a = 42;
  // ...
}

// valid:
function f(a){
  var _a = arguments.length < 1 ? 42 : a;
  // ...
}
```

**Comment by pnstickne** [comment]
These de-opts are engine-specific. A rule to target them should probably take that into account.

Anyway, _I_ would find it useful to have the first valid and invalid cases shown (but not the latter two cases). I do work on a game library and it can really make a difference..

**Comment by IanVS** [comment]
I wonder if it makes more sense to create a plugin for performance-related rules, particularly since much of it depends on JS engine implementations.

**Comment by nzakas** [comment]
I can't really discern a pattern between the invalid and valid examples. What's the key factor?

**Comment by zloirock** [comment]
@nzakas function arguments bound to `arguments` object. AFAIC it creates problems for optimization anywhere:
```js
function bar(args){ // <- we don't know anything about this function
  args[0] = 42;
}

function foo(a){
  console.log(a);
  bar(arguments);
  console.log(a); // equal first output? no, bound with arguments[0]
}

foo(21);
// => 21
// => 42
```
And the same if `a` will be changed. In V8 (and similar in other engines):
* Any calling function with `arguments` (exclude `.apply` - it's special case) deoptimize function.
* Reassign defined parameter if function also uses `arguments` deoptimize function - it cause changing `arguments` element.
* Safe for usage only access to `arguments` elements by index and `.length` property.

**Comment by nzakas** [comment]
Okay, I'm going to suggest you create a plugin that flags V8 optimization issues with rules. I don't think this belongs in the core because engine optimizations can change, and we don't want to target just one engine for a rule.

**Comment by drBenway** [comment]
would the following be a correct workaround?

var args = Array.from(arguments);
[MDN has a polyfill](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from)

**Comment by zloirock** [comment]
@drBenway `Array.from` will work, but `Array.from` will not be fast, only inline loop. Anyway, MDN polyfill incorrect because does not support iterables. Use [correct polyfill](https://github.com/zloirock/core-js#ecmascript-6-array).
